File: spark-event-ui/src/app/model/trackers/metrics/MetricValueHolder.ts
import {AccumUpdateValue, SparkListenerDriverAccumUpdates, SparkListenerTaskEnd} from '../../sparkevents/SparkEvent';
import {TaskTracker} from '../TaskTracker';
import {SparkPlanInfoTree} from '../SparkPlanNode';
import {Metrics} from '../../sparkevents/SparkPlanInfo';
import {AccumulableInfo} from '../../sparkevents/AccumulableInfo';
import {MetricDescription} from './MetricDescription';
import {MetricDescriptionRegistry} from './MetricDescriptionRegistry';
import {SparkPlanNode} from '../sql/SparkPlanNode';


/**
 * holder for metric value in SparkPlanInfoNode
 */
export class MetricValueHolder {
  readonly attachedNode: SparkPlanNode;

  readonly metric: Metrics;
  // name: string;
  // accumulatorId: number;
  // metricType: string;
  readonly metricDescription: MetricDescription;

  value: number = 0;

  updateCount: number = 0;
  updateCountPerDriver: number = 0;
  get updateCountNotDriver(): number { return this.updateCount - this.updateCountPerDriver; }
  readonly updatesPerJob: MetricValueUpdatesPerJob[] = [];


  get name(): string { return this.metric.name; }
  get accumulatorId(): number { return this.metric.accumulatorId; }
  get metricType(): string { return this.metric.metricType; }

  constructor(attachedNode: SparkPlanNode,
              metric: Metrics,
              previousTree?: SparkPlanInfoTree
  ) {
    this.attachedNode = attachedNode;
    this.metric = metric;
    this.metricDescription = MetricDescriptionRegistry.INSTANCE.resolve(metric.name, metric.metricType);
    attachedNode.tree.metricValueHolderById.set(this.accumulatorId, this);

    if (previousTree) {
      const previousValueHolder = previousTree.metricValueHolderById.get(this.accumulatorId);
      if (previousValueHolder) {
        this.value = previousValueHolder.value;
        this.updateCount = previousValueHolder.updateCount;
        this.updateCountPerDriver = previousValueHolder.updateCountPerDriver;
        previousValueHolder.updatesPerJob.forEach(prevPerJob => {
          this.updatesPerJob.push(prevPerJob.newUpdateCopyFor(this));
        });
      }
    }
  }

  isValueToShow(): boolean {
    // TODO..
    if (!this.value) {
      return false;
    }
    return true;
  }

  getAllValues(): number[] {
    return this.updatesPerJob.flatMap(
      perJob => perJob.updatesPerStages.flatMap(
        perStage => perStage.incrValues
      )
    );
  }

  onChildTaskEndMetricValueUpdate(task: TaskTracker, event: SparkListenerTaskEnd, acc: AccumulableInfo) {
    const jobId = task.stage.job.jobId;
    const foundIndex = this.updatesPerJob.findIndex(x => x.jobId === jobId);
    let perJob: MetricValueUpdatesPerJob;
    if (-1 === foundIndex) {
      perJob = new MetricValueUpdatesPerJob(this, jobId);
      this.updatesPerJob.push(perJob);
    } else {
      perJob = this.updatesPerJob[foundIndex];
    }
    perJob.onChildTaskEndMetricValueUpdate(task, event, acc);

    this.value = acc.value; // should be equivalent to this.value += acc.update;
    this.updateCount++;
  }


  onDriverAccumUpdate(event: SparkListenerDriverAccumUpdates, accUpdate: AccumUpdateValue) {
    this.value = accUpdate.value;
    this.updateCount++;
    this.updateCountPerDriver++;
  }


  updatesPerJobSummary(perStageDetails: boolean) {
    if (this.updatesPerJob.length === 0) {
      return '';
    } if (this.updatesPerJob.length === 1) {
      const perJob = this.updatesPerJob[0];
      return '(from job ' + perJob.jobId + perJob.optUpdatesPerStageSummary(perStageDetails) + ')';
    } else {
      return `(from ${this.updatesPerJob.length} jobs:` +
        ' ' + this.updatesPerJob.map(
          x => x.jobId + x.optUpdatesPerStageSummary(perStageDetails)
        ).join(', ') +
        ')';
    }
  }

  updatesSummary(showMetricDetailsUpdateCount: boolean,
                 showMetricDetailsPerDriver: boolean,
                 showMetricDetailsPerJob: boolean,
                 showMetricDetailsPerStage: boolean,
                 showMetricDetailsMinMaxValues: boolean,
                 showMetricDetailsValues: boolean) {
    if (this.updateCount === 0) {
      return "";
    }
    let res = '';
    if (showMetricDetailsUpdateCount) {
      if (this.updateCount === 1) {
        // res += ' updated once';
      } else {
        res += ` updated ${this.updateCount} time${this.updateCount > 1 ? 's' : ''}`;
      }
    }
    if (showMetricDetailsPerDriver) {
      if (this.updateCountPerDriver === 0) {
        //
      } else if (this.updateCountPerDriver === 1) {
        res += ' from driver';
      } else {
        res += ` ${this.updateCountPerDriver} from driver`;
      }
    }

    if (showMetricDetailsPerJob) {
      res += ' ' + this.updatesPerJobSummary(showMetricDetailsPerStage);
    }

    if (this.updateCountNotDriver > 1) {
      if (showMetricDetailsMinMaxValues) {
        const allValues = this.getAllValues();
        if (allValues.length === 0) {
          res += ' no values'; // ???? multiple update from Driver? values not recorded yet
        } else {
          const minValue = Math.min(...allValues);
          const maxValue = Math.max(...allValues);
          res += ` min: ${minValue}, max: ${maxValue}`;
        }
      }

      if (showMetricDetailsValues) {
        res += ' values [' + this.getAllValues().join(', ') + ']';
      }
    }
    return res;
  }
}

/**
 *
 */
export class MetricValueUpdatesPerJob {
  readonly updatesPerStages: MetricValueUpdatesPerJobStage[] = [];

  constructor(public readonly valueHolder: MetricValueHolder,
              public readonly jobId: number) {
  }

  onChildTaskEndMetricValueUpdate(task: TaskTracker, event: SparkListenerTaskEnd, acc: AccumulableInfo) {
    const stageId = task.stage.stageId;
    const foundIndex = this.updatesPerStages.findIndex(x => x.stageId === stageId);
    let perStage: MetricValueUpdatesPerJobStage;
    if (-1 === foundIndex) {
      perStage = new MetricValueUpdatesPerJobStage(this, stageId);
      this.updatesPerStages.push(perStage);
    } else {
      perStage = this.updatesPerStages[foundIndex];
    }
    perStage.onChildTaskEndMetricValueUpdate(task, event, acc);
  }

  optUpdatesPerStageSummary(enable: boolean): string {
    return enable ? ' ' + this.updatesPerStageSummary() : '';
  }

  updatesPerStageSummary() {
    if (this.updatesPerStages.length === 0) {
      return '';
    } if (this.updatesPerStages.length === 1) {
      const perStage = this.updatesPerStages[0];
      let res = 'stage: ' + perStage.stageId;
      return res;
    } else {
      return ` (${this.updatesPerStages.length} stages: ` +
        this.updatesPerStages.map(x => x.stageId).join(', ') +
        ')';
    }
  }

  newUpdateCopyFor(toValueHolder: MetricValueHolder): MetricValueUpdatesPerJob {
    const res = new MetricValueUpdatesPerJob(toValueHolder, this.jobId);
    this.updatesPerStages.forEach(prevPerStage => {
      res.updatesPerStages.push(prevPerStage.newUpdateCopyFor(this));
    });
    return res;
  }
}

/**
 *
 */
export class MetricValueUpdatesPerJobStage {
  count: number = 0;
  incrValuesSum: number = 0;
  readonly incrValues: number[] = [];
  readonly incrTimes: number[] = [];

  constructor(public readonly perJob: MetricValueUpdatesPerJob,
              public readonly stageId: number) {
  }

  onChildTaskEndMetricValueUpdate(task: TaskTracker, event: SparkListenerTaskEnd, acc: AccumulableInfo) {
    this.count++;
    this.incrValuesSum += acc.update;
    this.incrValues.push(acc.update);
    this.incrTimes.push(event.taskInfo.finishTime?.getTime() ?? 0);
  }

  newUpdateCopyFor(newPerJob: MetricValueUpdatesPerJob): MetricValueUpdatesPerJobStage {
    const res = new MetricValueUpdatesPerJobStage(newPerJob, this.stageId);
    res.count = this.count;
    res.incrValuesSum = this.incrValuesSum;
    res.incrValues.push(...this.incrValues);
    res.incrTimes.push(...this.incrTimes);
    return res;
  }
}

File: spark-event-gateway-srv/src/main/java/fr/an/spark/gateway/eventTrackers/sql/metrics/MetricValueHolder.java
package fr.an.spark.gateway.eventTrackers.sql.metrics;

import fr.an.spark.gateway.eventTrackers.model.TaskTracker;
import fr.an.spark.gateway.eventTrackers.sql.SparkPlanInfoTree;
import fr.an.spark.gateway.eventTrackers.sql.SparkPlanNode;
import fr.an.spark.gateway.eventlog.model.AccumulableInfo;
import fr.an.spark.gateway.eventlog.model.SparkEvent.AccumUpdateValue;
import fr.an.spark.gateway.eventlog.model.SparkEvent.SparkListenerDriverAccumUpdates;
import fr.an.spark.gateway.eventlog.model.SparkEvent.SparkListenerTaskEnd;
import fr.an.spark.gateway.eventlog.model.SparkPlanInfo.Metrics;
import fr.an.spark.gateway.utils.LsUtils;
import lombok.Getter;
import lombok.val;

import java.util.*;
import java.util.stream.Collectors;

/**
 *
 */
@Getter
public class MetricValueHolder {

    public final SparkPlanNode attachedNode;
    public final Metrics metric;
    public final MetricDescription metricDescription;

    public Object value;
    public int updateCount = 0;
    public int updateCountPerDriver = 0;
    public final List<MetricValueUpdatesPerJob> updatesPerJobs = new ArrayList<>();

    // -----------------------------------------------------------------------------------------------------------------

    public MetricValueHolder(SparkPlanNode attachedNode, Metrics metric, SparkPlanInfoTree previousTree) {
        this.attachedNode = attachedNode;
        this.metric = metric;
        this.metricDescription = MetricDescriptionRegistry.INSTANCE.resolve(metric.name, metric.metricType);
        attachedNode.tree.metricValueHolderById.put(getAccumulatorId(), this);

        if (previousTree != null) {
            MetricValueHolder previousValueHolder = previousTree.metricValueHolderById.get(getAccumulatorId());
            if (previousValueHolder != null) {
                this.value = previousValueHolder.value;
                this.updateCount = previousValueHolder.updateCount;
                this.updateCountPerDriver = previousValueHolder.updateCountPerDriver;
                for (MetricValueUpdatesPerJob prevPerJob : previousValueHolder.updatesPerJobs) {
                    this.updatesPerJobs.add(prevPerJob.newUpdateCopyFor(this));
                }
            }
        }
    }

    public String getName() { return metric.name; }
    public int getAccumulatorId() { return metric.accumulatorId; }
    public String getMetricType() { return metric.metricType; }
    public int getUpdateCountNotDriver() { return updateCount - updateCountPerDriver; }

    public boolean isValueToShow() {
        return value != null;
    }

    public List<Long> getAllValues() {
        return updatesPerJobs.stream()
                .flatMap(perJob -> perJob.updatesPerStages.stream()
                        .flatMap(perStage -> perStage.incrValues.stream()))
                .collect(Collectors.toList());
    }

    public void onChildTaskEndMetricValueUpdate(TaskTracker task, SparkListenerTaskEnd event, AccumulableInfo acc) {
        // TODO task -> in stage.. can be in several jobs
        for(val job: task.stage.jobs) {
            val jobId = job.jobId;
            MetricValueUpdatesPerJob perJob = LsUtils.findFirstOrAdd(updatesPerJobs, x -> x.jobId == jobId,
                    () -> new MetricValueUpdatesPerJob(this, jobId));
            perJob.onChildTaskEndMetricValueUpdate(task, event, acc);
        }

        this.value = acc.value;
        this.updateCount++;
    }

    public void onDriverAccumUpdate(SparkListenerDriverAccumUpdates event, AccumUpdateValue accUpdate) {
        this.value = accUpdate.value();
        this.updateCount++;
        this.updateCountPerDriver++;
    }

    public String updatesPerJobSummary(boolean perStageDetails) {
        if (updatesPerJobs.isEmpty()) {
            return "";
        }
        if (updatesPerJobs.size() == 1) {
            MetricValueUpdatesPerJob perJob = updatesPerJobs.get(0);
            return "(from job " + perJob.jobId + perJob.optUpdatesPerStageSummary(perStageDetails) + ")";
        } else {
            return "(from " + updatesPerJobs.size() + " jobs: " +
                    updatesPerJobs.stream()
                            .map(x -> x.jobId + x.optUpdatesPerStageSummary(perStageDetails))
                            .collect(Collectors.joining(", ")) +
                    ")";
        }
    }

    public String updatesSummary(boolean showMetricDetailsUpdateCount,
                                 boolean showMetricDetailsPerDriver,
                                 boolean showMetricDetailsPerJob,
                                 boolean showMetricDetailsPerStage,
                                 boolean showMetricDetailsMinMaxValues,
                                 boolean showMetricDetailsValues) {
        if (updateCount == 0) {
            return "";
        }
        StringBuilder res = new StringBuilder();
        if (showMetricDetailsUpdateCount) {
            if (updateCount != 1) {
                res.append(" updated ").append(updateCount).append(" time").append(updateCount > 1 ? "s" : "");
            }
        }
        if (showMetricDetailsPerDriver) {
            if (updateCountPerDriver == 1) {
                res.append(" from driver");
            } else if (updateCountPerDriver > 1) {
                res.append(" ").append(updateCountPerDriver).append(" from driver");
            }
        }
        if (showMetricDetailsPerJob) {
            res.append(" ").append(updatesPerJobSummary(showMetricDetailsPerStage));
        }
        if (getUpdateCountNotDriver() > 1) {
            if (showMetricDetailsMinMaxValues) {
                List<Long> allValues = getAllValues();
                if (allValues.isEmpty()) {
                    res.append(" no values");
                } else {
                    long minValue = Collections.min(allValues);
                    long maxValue = Collections.max(allValues);
                    res.append(" min: ").append(minValue).append(", max: ").append(maxValue);
                }
            }
            if (showMetricDetailsValues) {
                res.append(" values [")
                        .append(getAllValues().stream().map(String::valueOf).collect(Collectors.joining(", ")))
                        .append("]");
            }
        }
        return res.toString();
    }

    // -----------------------------------------------------------------------------------------------------------------

    public static class MetricValueUpdatesPerJob {
        public final List<MetricValueUpdatesPerJobStage> updatesPerStages = new ArrayList<>();
        public final MetricValueHolder valueHolder;
        public final int jobId;

        public MetricValueUpdatesPerJob(MetricValueHolder valueHolder, int jobId) {
            this.valueHolder = valueHolder;
            this.jobId = jobId;
        }

        public void onChildTaskEndMetricValueUpdate(TaskTracker task, SparkListenerTaskEnd event, AccumulableInfo acc) {
            int stageId = task.stageId;
            MetricValueUpdatesPerJobStage perStage = LsUtils.findFirst(updatesPerStages, x -> x.stageId == stageId);
            if (perStage == null) {
                perStage = new MetricValueUpdatesPerJobStage(this, stageId);
                updatesPerStages.add(perStage);
            }
            perStage.onChildTaskEndMetricValueUpdate(task, event, acc);
        }

        public String optUpdatesPerStageSummary(boolean enable) {
            return enable ? " " + updatesPerStageSummary() : "";
        }

        public String updatesPerStageSummary() {
            if (updatesPerStages.isEmpty()) {
                return "";
            }
            if (updatesPerStages.size() == 1) {
                MetricValueUpdatesPerJobStage perStage = updatesPerStages.get(0);
                return "stage: " + perStage.stageId;
            } else {
                return " (" + updatesPerStages.size() + " stages: " +
                        updatesPerStages.stream().map(x -> String.valueOf(x.stageId)).collect(Collectors.joining(", ")) +
                        ")";
            }
        }

        public MetricValueUpdatesPerJob newUpdateCopyFor(MetricValueHolder toValueHolder) {
            MetricValueUpdatesPerJob res = new MetricValueUpdatesPerJob(toValueHolder, this.jobId);
            for (MetricValueUpdatesPerJobStage prevPerStage : this.updatesPerStages) {
                res.updatesPerStages.add(prevPerStage.newUpdateCopyFor(res));
            }
            return res;
        }
    }

    public static class MetricValueUpdatesPerJobStage {
        public int count = 0;
        public long incrValuesSum = 0;
        public final List<Long> incrValues = new ArrayList<>();
        public final List<Long> incrTimes = new ArrayList<>();
        public final MetricValueUpdatesPerJob perJob;
        public final int stageId;

        public MetricValueUpdatesPerJobStage(MetricValueUpdatesPerJob perJob, int stageId) {
            this.perJob = perJob;
            this.stageId = stageId;
        }

        public void onChildTaskEndMetricValueUpdate(TaskTracker task, SparkListenerTaskEnd event, AccumulableInfo acc) {
            this.count++;
            Object accValue = acc.update;
            if (accValue instanceof Number accValueNum) {
                val accValueLong = accValueNum.longValue();
                this.incrValuesSum += accValueLong;
                this.incrValues.add(accValueLong);
            } else {
                // should not occur
            }
            this.incrTimes.add(event.taskInfo.finishTime);
        }

        public MetricValueUpdatesPerJobStage newUpdateCopyFor(MetricValueUpdatesPerJob newPerJob) {
            MetricValueUpdatesPerJobStage res = new MetricValueUpdatesPerJobStage(newPerJob, this.stageId);
            res.count = this.count;
            res.incrValuesSum = this.incrValuesSum;
            res.incrValues.addAll(this.incrValues);
            res.incrTimes.addAll(this.incrTimes);
            return res;
        }
    }
}
